crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -547,50 +583,55 @@ impl<'w, T: Fetch<'w>> Fetch<'w> for OptionFetch<T> {
 
 /// Flags on component `T` that happened since the start of the frame.
 #[derive(Clone)]
-pub struct Flags<T: Component> {
-    flags: ComponentFlags,
+pub struct Counters<T: Component> {
+    component_counters: ComponentCounters,
+    system_counter: Option<u32>,
+    global_system_counter: u32,
     marker: PhantomData<T>,
 }
-impl<T: Component> std::fmt::Debug for Flags<T> {
+impl<T: Component> std::fmt::Debug for Counters<T> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Flags")
-            .field("added", &self.added())
-            .field("mutated", &self.mutated())
+        f.debug_struct("Counters")
+            .field("component_counters", &self.component_counters)
+            .field("system_counter", &self.system_counter)
+            .field("global_system_counter", &self.global_system_counter)
             .finish()
     }
 }
 
-impl<T: Component> Flags<T> {
-    /// Has this component been added since the start of the frame.
-    pub fn added(&self) -> bool {
-        self.flags.contains(ComponentFlags::ADDED)
+impl<T: Component> Counters<T> {
+    /// Has this component been added since the last execution of this system.
+    pub fn is_added(&self) -> bool {
+        self.component_counters
+            .is_added(self.system_counter, self.global_system_counter)
     }
 
-    /// Has this component been mutated since the start of the frame.
-    pub fn mutated(&self) -> bool {
-        self.flags.contains(ComponentFlags::MUTATED)
+    /// Has this component been mutated since the last execution of this system.
+    pub fn is_mutated(&self) -> bool {
```

Comment:
I thought we were going to remove `Mutated` in general?

Reply:
I'll remove it cleanly in a separate commit so that it's easy to revert if needed